libs/core/likwid/include/hpx/likwid/likwid_executor.hpp

```diff
@@ -0,0 +1,230 @@
+//  Copyright (c) 2022 Srinivas Yadav
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <likwid.h>
+#include <hpx/execution/execution.hpp>
+
+namespace hpx { namespace execution {
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename BaseExecutor>
+    class likwid_executor
+    {
+    private:
+        struct on_exit
+        {
+            explicit on_exit(likwid_executor const& exec)
+              : exec_(exec)
+            {
+                exec_.on_start_();
+            }
+
+            ~on_exit()
+            {
+                exec_.on_stop_();
+            }
+
+            likwid_executor const& exec_;
+        };
+
+        template <typename F>
+        struct hook_wrapper
+        {
+            template <typename... Ts>
+            decltype(auto) operator()(Ts&&... ts)
+            {
+                on_exit _{exec_};
+                return hpx::invoke(f_, std::forward<Ts>(ts)...);
+            }
+
+            likwid_executor const& exec_;
+            F f_;
+        };
+
+    public:
+        using execution_category = typename 
+            hpx::parallel::execution::executor_execution_category<BaseExecutor>::type;
+        using executor_parameters_type =
+            typename hpx::parallel::execution::executor_parameters_type<BaseExecutor>::type;
+
+        likwid_executor(
+            BaseExecutor& exec, std::string name)
+          : exec_(exec)
+        {
+            on_start_ = [name]() {likwid_markerStartRegion(name.data());};
+            on_stop_ = [name]() {likwid_markerStopRegion(name.data());};
```

Comment:
Do we have to copy the `name` twice?

Reply:
Also, shouldn't this use `name.c_str()` instead (i.e. ensure the sequence of bytes is zero-terminated)?